src/chainstate/stacks/boot/pox_2_tests.rs

```diff
@@ -0,0 +1,419 @@
+use vm::database::*;
+use vm::errors::{
+    CheckErrors, Error, IncomparableError, InterpreterError, InterpreterResult, RuntimeErrorType,
+};
+use vm::eval;
+use vm::representations::SymbolicExpression;
+use vm::tests::{execute, is_committed, is_err_code, symbols_from_values};
+use vm::types::Value::Response;
+use vm::types::{
+    OptionalData, PrincipalData, QualifiedContractIdentifier, ResponseData, StandardPrincipalData,
+    TupleData, TupleTypeSignature, TypeSignature, Value, NONE,
+};
+
+use crate::{
+    burnchains::Burnchain,
+    chainstate::{
+        burn::db::sortdb::SortitionDB,
+        stacks::{events::TransactionOrigin, miner::test::make_coinbase},
+    },
+    clarity_vm::{clarity::ClarityBlockConnection, database::marf::WritableMarfStore},
+    net::test::TestEventObserver,
+    util::boot::boot_code_id,
+};
+use types::chainstate::{
+    BlockHeaderHash, BurnchainHeaderHash, StacksAddress, StacksBlockId, VRFSeed,
+};
+use types::proof::{ClarityMarfTrieId, TrieMerkleProof};
+
+use clarity_vm::clarity::Error as ClarityError;
+
+use super::test::*;
+
+const USTX_PER_HOLDER: u128 = 1_000_000;
+
+/// In this test case, two Stackers, Alice and Bob stack and interact with the
+///  PoX v1 contract and PoX v2 contract across the epoch transition.
+///
+/// Alice: stacks via PoX v1 for 4 cycles. The third of these cycles occurs after
+///        the PoX v1 -> v2 transition, and so Alice gets "early unlocked".
+///        After the early unlock, Alice re-stacks in PoX v2
+///        Alice tries to stack again via PoX v1, which is allowed by the contract,
+///        but forbidden by the VM (because PoX has transitioned to v2)
+/// Bob:   stacks via PoX v2 for 6 cycles. He attempted to stack via PoX v1 as well,
+///        but is forbidden because he has already placed an account lock via PoX v2.
+///
+#[test]
+fn test_simple_pox_lockup_transition_pox_2() {
+    let AUTO_UNLOCK_HT = 12;
+    let mut burnchain = Burnchain::default_unittest(0, &BurnchainHeaderHash::zero());
+    burnchain.pox_constants.reward_cycle_length = 5;
+    burnchain.pox_constants.prepare_length = 2;
+    burnchain.pox_constants.anchor_threshold = 1;
+    burnchain.pox_constants.v1_unlock_height = AUTO_UNLOCK_HT + 25;
+
+    let first_v2_cycle = burnchain
+        .block_height_to_reward_cycle(burnchain.pox_constants.v1_unlock_height as u64)
+        .unwrap()
+        + 1;
+
+    eprintln!("First v2 cycle = {}", first_v2_cycle);
+
+    let epochs = StacksEpoch::all(0, 25 + 10);
+
+    let observer = TestEventObserver::new();
+
+    let (mut peer, mut keys) = instantiate_pox_peer_with_epoch(
+        &burnchain,
+        "test_simple_pox_lockup_transition_pox_2",
+        6002,
+        Some(epochs.clone()),
+        Some(&observer),
+    );
+
+    let num_blocks = 35;
+
+    let alice = keys.pop().unwrap();
+    let bob = keys.pop().unwrap();
+    let charlie = keys.pop().unwrap();
+
+    let mut alice_reward_cycle = 0;
+
+    for tenure_id in 0u32..num_blocks {
```

Comment:
I think this test would be easier to understand if it were not written in a for loop. Instead, it could be written as an alternating series of 1) execute a tenure, 2) check the relevant invariants after that tenure.

E.g.:

// Do tenure 1 by running code "code for tenure 1"
// Test what is true after tenure 1

// Do tenure 1 by running code "code for tenure 2"
// Test what is true after tenure 2

... etc


If it were done this way, without a loop, and with a more declarative syntax, I think it would have the following benefits:
1) Easier to read - locality: Right now, the logic that changes the state is not near the logic that checks the state. So, I have to scroll back and forth and try to figure out what is being checked after each test. I think it would be easier to read if you could organize the code as a sequence of statements, where what is being checked is checked right after the relevant change.
2) Easier to read - no thinking: This test has a lot of logic to understand what is tested when. It's confusing to have to do so much logic to read a test. The whole point of the test is to test the logic of the code under test, so it's harder to know what the test result should be if the test also has a lot of logic.
2)More fool-proof test: With so much logic going into the test, we run the risk of the test-logic having a bug.

Reply:
Yes, I agree with that. I'll unroll the test.

Reply:
Unrolled in 6c69759 -- I *think* it improved the clarity of the test.